include/nil/actor/zk/commitments/detail/polynomial/proof_of_work.hpp

```diff
@@ -83,6 +87,96 @@ namespace nil {
                         return ((result & mask) == 0);
                     }
                 };
+
+                // Note that the interface here is slightly different from the one above:
+                // amount of bits for grinding instead of the mask.
+                // This was done because the actual mask is applied to the high bits instead of the low bits
+                // which makes manually setting the mask error-prone.
+                template<typename TranscriptHashType, typename FieldType, std::uint8_t GrindingBits=16>
+                class field_proof_of_work {
+                public:
+                    using transcript_hash_type = TranscriptHashType;
+                    using transcript_type = transcript::fiat_shamir_heuristic_sequential<transcript_hash_type>;
+                    using value_type = typename FieldType::value_type;
+                    using integral_type = typename FieldType::integral_type;
+
+                    constexpr static const integral_type mask =
+                        (GrindingBits > 0 ?
+                            ((integral_type(2) << GrindingBits - 1) - 1) << (FieldType::modulus_bits - GrindingBits)
+                            : 0);
+
+                    static inline boost::property_tree::ptree get_params() {
+                        boost::property_tree::ptree params;
+                        params.put("mask", mask);
+                        return params;
+                    }
+
+                    static inline value_type generate(transcript_type &transcript, 
+                        nil::crypto3::random::algebraic_engine<FieldType> random_engine) {
+
+                        value_type pow_seed = random_engine();
+                        value_type init_seed = pow_seed;
+                        std::size_t per_core = 4*65536;
```

Comment:
This must probably depend on GrindingBits?

Reply:
If I set GrindingBits = 10, how will this work?

Reply:
This will launch `smp::count` workers, each with `1<<18` values to check. First one who finds, sets `challenge_found` to true and returns the value. Others stop once they see `challenge_found` is set.

Is there a guarantee, that 10 bits could be ground with `1<<10` values?
I suppose one of the workers will finish when ~ `1<<10` values are checked and there will be no other round.

Reply:
In this case, can we get rid of the value "per_core"? Maybe each core will iterate over the numbers which give a given reminder on the smp::count?

Like core #i will iterate over numbers i + N * smp::count until it finds something or another core stops it?